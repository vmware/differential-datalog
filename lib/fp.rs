use crate::int::*;
use crate::uint::*;
use num::bigint::BigInt;
use num_traits::cast::FromPrimitive;
use ordered_float::OrderedFloat;

pub fn floor_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.floor())
}

pub fn ceil_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.ceil())
}

pub fn round_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.round())
}

pub fn trunc_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.trunc())
}

pub fn fract_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.fract())
}

pub fn abs_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.abs())
}

pub fn signum_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.signum())
}

pub fn sqrt_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.sqrt())
}

pub fn exp_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.exp())
}

pub fn exp2_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.exp2())
}

pub fn ln_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.ln())
}

pub fn log2_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.log2())
}

pub fn log10_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.log10())
}

pub fn cbrt_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.cbrt())
}

pub fn sin_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.sin())
}

pub fn cos_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.cos())
}

pub fn tan_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.tan())
}

pub fn asin_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.asin())
}

pub fn acos_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.acos())
}

pub fn atan_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.atan())
}

pub fn atan2_f(f: &OrderedFloat<f32>, other: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.atan2(**other))
}

pub fn sinh_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.sinh())
}

pub fn cosh_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.cosh())
}

pub fn tanh_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.tanh())
}

pub fn asinh_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.asinh())
}

pub fn acosh_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.acosh())
}

pub fn atanh_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.atanh())
}

pub fn recip_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.recip())
}

pub fn to_degrees_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.to_degrees())
}

pub fn to_radians_f(f: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(f.to_radians())
}

pub fn is_nan_f(f: &OrderedFloat<f32>) -> bool {
    f.is_nan()
}

pub fn is_infinite_f(f: &OrderedFloat<f32>) -> bool {
    f.is_infinite()
}

pub fn is_finite_f(f: &OrderedFloat<f32>) -> bool {
    f.is_normal()
}

pub fn mul_add_f(
    a: &OrderedFloat<f32>,
    b: &OrderedFloat<f32>,
    c: &OrderedFloat<f32>,
) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(a.mul_add(**b, **c))
}

pub fn powi_f(n: &OrderedFloat<f32>, exp: &i32) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(n.powi(*exp))
}

pub fn powf_f(n: &OrderedFloat<f32>, exp: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(n.powf(**exp))
}

pub fn log_f(n: &OrderedFloat<f32>, base: &OrderedFloat<f32>) -> OrderedFloat<f32> {
    OrderedFloat::<f32>(n.log(**base))
}

pub fn nan_f() -> OrderedFloat<f32> {
    OrderedFloat::<f32>(::std::f32::NAN)
}

//-------------------------------

pub fn floor_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.floor())
}

pub fn ceil_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.ceil())
}

pub fn round_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.round())
}

pub fn trunc_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.trunc())
}

pub fn fract_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.fract())
}

pub fn abs_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.abs())
}

pub fn signum_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.signum())
}

pub fn sqrt_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.sqrt())
}

pub fn exp_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.exp())
}

pub fn exp2_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.exp2())
}

pub fn ln_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.ln())
}

pub fn log2_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.log2())
}

pub fn log10_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.log10())
}

pub fn cbrt_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.cbrt())
}

pub fn sin_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.sin())
}

pub fn cos_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.cos())
}

pub fn tan_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.tan())
}

pub fn asin_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.asin())
}

pub fn acos_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.acos())
}

pub fn atan_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.atan())
}

pub fn atan2_d(f: &OrderedFloat<f64>, other: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.atan2(**other))
}

pub fn sinh_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.sinh())
}

pub fn cosh_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.cosh())
}

pub fn tanh_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.tanh())
}

pub fn asinh_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.asinh())
}

pub fn acosh_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.acosh())
}

pub fn atanh_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.atanh())
}

pub fn recip_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.recip())
}

pub fn to_degrees_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.to_degrees())
}

pub fn to_radians_d(f: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(f.to_radians())
}

pub fn is_nan_d(f: &OrderedFloat<f64>) -> bool {
    f.is_nan()
}

pub fn is_infinite_d(f: &OrderedFloat<f64>) -> bool {
    f.is_infinite()
}

pub fn is_finite_d(f: &OrderedFloat<f64>) -> bool {
    f.is_normal()
}

pub fn mul_add_d(
    a: &OrderedFloat<f64>,
    b: &OrderedFloat<f64>,
    c: &OrderedFloat<f64>,
) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(a.mul_add(**b, **c))
}

pub fn powi_d(n: &OrderedFloat<f64>, exp: &i32) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(n.powi(*exp))
}

pub fn powf_d(n: &OrderedFloat<f64>, exp: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(n.powf(**exp))
}

pub fn log_d(n: &OrderedFloat<f64>, base: &OrderedFloat<f64>) -> OrderedFloat<f64> {
    OrderedFloat::<f64>(n.log(**base))
}

pub fn nan_d() -> OrderedFloat<f64> {
    OrderedFloat::<f64>(::std::f64::NAN)
}

pub fn int_from_f(v: &OrderedFloat<f32>) -> crate::ddlog_std::Option<Int> {
    match (BigInt::from_f32(**v)) {
        None => crate::ddlog_std::Option::None,
        Some(x) => crate::ddlog_std::Option::Some {
            x: Int::from_bigint(x),
        },
    }
}

pub fn int_from_d(v: &OrderedFloat<f64>) -> crate::ddlog_std::Option<Int> {
    match (BigInt::from_f64(**v)) {
        None => crate::ddlog_std::Option::None,
        Some(x) => crate::ddlog_std::Option::Some {
            x: Int::from_bigint(x),
        },
    }
}

pub fn parse_f(s: &String) -> crate::ddlog_std::Result<OrderedFloat<f32>, String> {
    match (s.parse::<f32>()) {
        Ok(res) => crate::ddlog_std::Result::Ok {
            res: OrderedFloat::<f32>(res),
        },
        Err(err) => crate::ddlog_std::Result::Err {
            err: format!("{}", err),
        },
    }
}

pub fn parse_d(s: &String) -> crate::ddlog_std::Result<OrderedFloat<f64>, String> {
    match (s.parse::<f64>()) {
        Ok(res) => crate::ddlog_std::Result::Ok {
            res: OrderedFloat::<f64>(res),
        },
        Err(err) => crate::ddlog_std::Result::Err {
            err: format!("{}", err),
        },
    }
}
