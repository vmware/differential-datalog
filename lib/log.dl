/*
 * Logging API.
 *
 * This API has two facets.  First, it provides the `log()` function that can
 * be invoked from a DDlog rule's Inspect operator to provide more visibility
 * into the rule's evaluation.  In addition to the usual log level and log
 * message arguments, this function takes a `module` argument, which identifies
 * the module or subsystem that this log message is associated with.
 *
 * Second, it provides a Rust configuration API that can only be invoked by the
 * "host" program to set log level for individual modules and to supply a
 * logging callback for each module (see `rust/template/types/log.rs`).  This
 * way, different modules can produce separate log streams or even use different
 * logging mechanisms.
 *
 * NOTE: Internally, this library maintains a global module-to-log_callback
 * map shared by all DDlog instances running in the same address space.  The host
 * program is responsible for maintaining this mapping via the Rust API in
 * `log.rs`.
 *
 * Other related files:
 * `log4j.dl` - specializes the logging API for using log4j loggers
 * `rust/template/types/log.rs` - Rust logging API
 * `ddlog_log.h` - matching C bindings
 * `DDlogAPI.java (DDlogAPI.log_set_callback method)` - Java API to control
 *  logging.
 *
 * === Example ===
 *
 * In DDlog:
 * ```
 * // Declare module id's for logging purposes
 * function mod_SPAN_UUID(): log.module_t = 100
 *
 * // Use log() function in a rule
 * Span(entity, tn) :- Binding(entity, tn),
 *                     Inspect log(mod_SPAN_UUID(), 5, "Span(entity,tn) :- Binding(.entity=${deref(entity)},.tn=${deref(tn)}))").
 * ```
 *
 * In Java:
 * ```
 * // Module id declarations must match DDlog values:
 * static int MOD_SPAN_UUID1 = 100;
 * // Set logging callback and log level from Java
 * DDlogAPI.log_set_callback(
 *     MOD_SPAN_UUID1,
 *     (msg, level) -> System.err.println("Log msg from module1 (" + level + "): " + msg),
 *     5);
 * ```
 * ===============
 *
 */

/* `module_t` type identifies the subsystem or component for logging purposes.
 */
typedef module_t = signed<32>

typedef log_level_t = signed<32>

/*
 * Write to a log stream
 *
 * `module` - Identifies the subsystem writing to the log.  A log record is only
 *            produced if logging has been enabled for this subsystem (see
 *            `log.rs`)
 * `level`  - log level
 * `msg`    - error message (UTF-8)
 *
 * This function returns a `()` so that it can be used as part of an expression
 * of an Inspect operator.
 */
#[has_side_effects]
extern function log(module: module_t, level: log_level_t, msg: string): ()
