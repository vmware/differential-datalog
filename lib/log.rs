use std::collections;
use std::ffi;
use std::sync;

type log_callback_t = Box<dyn Fn(log_log_level_t, &str) + Send + Sync>;

lazy_static! {
    /* Logger configuration for each module consists of the maximal enabled
     * log level (messages above this level are ignored) and callback.
     */
    static ref LOG_CONFIG: sync::RwLock<collections::HashMap<log_module_t, (log_callback_t, log_log_level_t)>> = {
        sync::RwLock::new(collections::HashMap::new())
    };
}

/*
 * Logging API exposed to the DDlog program.
 * (see detailed documentation in `log.dl`)
 */
pub fn log_log(module: &log_module_t, level: &log_log_level_t, msg: &String) -> bool {
    if let Some((cb, current_level)) = LOG_CONFIG.read().unwrap().get(&module) {
        if *level <= *current_level {
            cb(*level, msg.as_str());
        }
    };
    true
}

/*
 * Configuration API
 * (detailed documentation in `ddlog_log.h`)
 *
 * `cb = None` - disables logging for the given module.
 *
 * NOTE: we set callback and log level simultaneously.  A more flexible API
 * would allow changing log level without changing the callback.
 */
pub fn log_set_callback(
    module: log_module_t,
    cb: Option<log_callback_t>,
    max_level: log_log_level_t,
) {
    match cb {
        Some(cb) => {
            LOG_CONFIG.write().unwrap().insert(module, (cb, max_level));
        }
        None => {
            LOG_CONFIG.write().unwrap().remove(&module);
        }
    }
}

/*
 * C bindings for the config API
 */
#[no_mangle]
pub unsafe extern "C" fn ddlog_log_set_callback(
    module: raw::c_int,
    cb: Option<extern "C" fn(arg: libc::uintptr_t, level: raw::c_int, msg: *const raw::c_char)>,
    cb_arg: libc::uintptr_t,
    max_level: raw::c_int,
) {
    match cb {
        Some(cb) => log_set_callback(
            module as log_module_t,
            Some(Box::new(move |level, msg| {
                cb(
                    cb_arg,
                    level as raw::c_int,
                    ffi::CString::new(msg).unwrap_or_default().as_ptr(),
                )
            })),
            max_level as log_log_level_t,
        ),
        None => log_set_callback(module as log_module_t, None, max_level as log_log_level_t),
    }
}
