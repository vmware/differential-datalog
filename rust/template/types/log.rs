#![allow(
    clippy::ptr_arg,
    clippy::trivially_copy_pass_by_ref
)]

use lazy_static::lazy_static;
use std::collections;
use std::ffi;
use std::os::raw;
use std::sync;

type log_callback_t = Box<dyn Fn(i32, &str) + Send + Sync>;

struct LogConfig {
    default_callback: Option<log_callback_t>,
    default_level: i32,
    mod_callbacks: collections::HashMap<i32, (log_callback_t, i32)>,
}

impl LogConfig {
    fn new() -> LogConfig {
        LogConfig {
            default_callback: None,
            default_level: std::i32::MAX,
            mod_callbacks: collections::HashMap::new(),
        }
    }
}

lazy_static! {
    /* Logger configuration for each module consists of the maximal enabled
     * log level (messages above this level are ignored) and callback.
     */
    static ref LOG_CONFIG: sync::RwLock<LogConfig> = {
        sync::RwLock::new(LogConfig::new())
    };
}

/*
 * Logging API exposed to the DDlog program.
 * (see detailed documentation in `log.dl`)
 */
pub fn log_log(module: &i32, level: &i32, msg: &String) {
    let cfg = LOG_CONFIG.read().unwrap();
    if let Some((cb, current_level)) = cfg.mod_callbacks.get(&module) {
        if *level <= *current_level {
            cb(*level, msg.as_str());
        }
    } else if *level <= cfg.default_level && cfg.default_callback.is_some() {
        (cfg.default_callback.as_ref().unwrap())(*level, msg.as_str());
    }
}

/*
 * Configuration API
 * (detailed documentation in `ddlog_log.h`)
 */

/*
 * `cb = None` - disables logging for the given module.
 *
 * NOTE: we set callback and log level simultaneously.  A more flexible API
 * would allow changing log level without changing the callback.
 */
pub fn log_set_callback(module: i32, cb: Option<log_callback_t>, max_level: i32) {
    let mut cfg = LOG_CONFIG.write().unwrap();
    match cb {
        Some(cb) => {
            cfg.mod_callbacks.insert(module, (cb, max_level));
        }
        None => {
            cfg.mod_callbacks.remove(&module);
        }
    }
}

/*
 * Set default callback and log level for modules that were not configured
 * via `log_set_callback`.
 */
pub fn log_set_default_callback(cb: Option<log_callback_t>, max_level: i32) {
    let mut cfg = LOG_CONFIG.write().unwrap();
    cfg.default_callback = cb;
    cfg.default_level = max_level;
}

/*
 * C bindings for the config API
 */
#[no_mangle]
pub unsafe extern "C" fn ddlog_log_set_callback(
    module: raw::c_int,
    cb: Option<extern "C" fn(arg: libc::uintptr_t, level: raw::c_int, msg: *const raw::c_char)>,
    cb_arg: libc::uintptr_t,
    max_level: raw::c_int,
) {
    match cb {
        Some(cb) => log_set_callback(
            module as i32,
            Some(Box::new(move |level, msg| {
                cb(
                    cb_arg,
                    level as raw::c_int,
                    ffi::CString::new(msg).unwrap_or_default().as_ptr(),
                )
            })),
            max_level as i32,
        ),
        None => log_set_callback(module as i32, None, max_level as i32),
    }
}

#[no_mangle]
pub unsafe extern "C" fn ddlog_log_set_default_callback(
    cb: Option<extern "C" fn(arg: libc::uintptr_t, level: raw::c_int, msg: *const raw::c_char)>,
    cb_arg: libc::uintptr_t,
    max_level: raw::c_int,
) {
    match cb {
        Some(cb) => log_set_default_callback(
            Some(Box::new(move |level, msg| {
                cb(
                    cb_arg,
                    level as raw::c_int,
                    ffi::CString::new(msg).unwrap_or_default().as_ptr(),
                )
            })),
            max_level as i32,
        ),
        None => log_set_default_callback(None, max_level as i32),
    }
}
