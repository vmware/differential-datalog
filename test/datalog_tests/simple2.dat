dump RFloatToInt;

start;

insert Arrng1(TArrng2{true, TArrng1{true, 1000}}, 10),
insert Arrng2(TArrng2{true, TArrng1{true, 5}}, 1000, true),

insert TArrng1[(TArrng2{true, TArrng1{true, 1000}}, 10)],
insert TArrng2[(TArrng2{true, TArrng1{true, 5}}, 1000)],

commit dump_changes;

dump FuncTest;

start;

insert Doubles([1.0, 2.0, 3.5]),

commit dump_changes;

dump InspectSimpleSum;

start;

insert InputTuples(1, 1);
insert InputTuples(1, 2);

commit dump_changes;

dump FilteredRelation;

start;

insert TestRelation(1, 1);
insert TestRelation(1, 2);
insert TestRelation(6, 2);
insert TestRelation(6, 3);
insert TestRelation(3, 4);
insert TestRelation(6, 5);

commit dump_changes;

start;

delete TestRelation(3, 4);
delete TestRelation(6, 2);

commit dump_changes;

dump OutputInspectNot;

start;

insert InputInspectNot2(1);
insert InputInspectNot2(2);
insert InputInspectNot2(3);

commit dump_changes;

start;

insert InputInspectNot1(1, 1);
insert InputInspectNot1(2, 2);
insert InputInspectNot1(3, 3);
insert InputInspectNot1(4, 4);

commit dump_changes;

start;

delete InputInspectNot2(2);

commit dump_changes;
