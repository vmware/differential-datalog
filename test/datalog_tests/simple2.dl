import fp
import log

/* Test `arrangeInput` logic in Compile.hs */

typedef TArrng1 = TArrng1{
    f1: bool,
    f2: bigint
}

typedef TArrng2 = TArrng2{
    f1: bool,
    f2: Ref<Ref<TArrng1>>
}

input relation Arrng1(
    f1: Ref<Ref<Ref<TArrng2>>>,
    f2: bigint
)

input relation Arrng2(
    f1: Ref<TArrng2>,
    f2: bigint,
    f3: bool
)

output relation Arrng1Arrng2(x: bigint)

Arrng1Arrng2(x.f2.f2) :-
    Arrng1(f1, f2),
    Arrng2(x, f1.f2.f2, _).

output relation Arrng1Arrng2_2(x: bigint)

Arrng1Arrng2_2(x.f2.f2) :-
    Arrng1(f1, f2),
    Arrng2(x, f1.f2.f2, f1.f2.f1).

input relation TArrng1[(Ref<Ref<Ref<TArrng2>>>, bigint)]

input relation TArrng2[(Ref<TArrng2>, bigint)]

output relation TArrng1Arrng2(x: bigint)

TArrng1Arrng2(x.f2.f2) :-
    TArrng1[t],
    TArrng2[(x, t.0.f2.f2)].


function ftoi_(l: double): signed<32> = {
    match (int_from_d(l)) {
        None -> 0,
        Some{x} -> {
            x as signed<32>
        }
    }
}

relation RFtoIDummy(x: signed<32>)
RFtoIDummy(0).
output relation RFloatToInt(_x:signed<32>)
RFloatToInt(_x) :- RFtoIDummy(0), var _x = ftoi_((- (64'f333.36: double))).

/* See #596 */
input relation Ints(x: bigint)
output relation SomeInts(x: Option<bigint>)

SomeInts(y) :-
    Ints(x),
    var y = Some{x},
    SomeInts(y).

/* See #592 */
function agg_avg_double_N(aggregate: Option<(double, double)>, item: Option<double>):
    Option<(double, double)> =
    match ((aggregate, item)) {
        (None, None)       -> None,
        (None, Some{x})    -> Some{(x, 1)},
        (Some{x}, None)    -> Some{x},
        (Some{(sum,ct)}, Some{y}) -> Some{(sum + y, ct + 1)}
    }


/* See #603. New function syntax avoids ambiguity. */
output relation &FuncTest(x: string)
function myfunc(x: string): string { x }
&FuncTest("foo").

output relation SumsOfDoubles(x: double, y: double, sum: double)
input relation Doubles(xs: Vec<double>)

SumsOfDoubles(x, y, z) :-
    Doubles(xs),
    var x = FlatMap(xs),
    Doubles(ys),
    var y = FlatMap(ys),
    var z = x + y.

/* Attempt to confuse statics */

function weird_zero(x: 'A): usize {
    var empty_vec: Vec<'A> = vec_empty();
    vec_len(empty_vec)
}

function zero_test(): usize {
    weird_zero(32'd0)
}
