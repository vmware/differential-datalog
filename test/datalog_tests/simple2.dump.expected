RFloatToInt{._x = -333}
Arrng1Arrng2:
Arrng1Arrng2{.x = 5}: +1
Arrng1Arrng2_2:
Arrng1Arrng2_2{.x = 5}: +1
TArrng1Arrng2:
TArrng1Arrng2{.x = 5}: +1
FuncTest{.x = "foo"}
SumsOfDoubles:
SumsOfDoubles{.x = 1, .y = 1, .sum = 2}: +1
SumsOfDoubles{.x = 1, .y = 2, .sum = 3}: +1
SumsOfDoubles{.x = 1, .y = 3.5, .sum = 4.5}: +1
SumsOfDoubles{.x = 2, .y = 1, .sum = 3}: +1
SumsOfDoubles{.x = 2, .y = 2, .sum = 4}: +1
SumsOfDoubles{.x = 2, .y = 3.5, .sum = 5.5}: +1
SumsOfDoubles{.x = 3.5, .y = 1, .sum = 4.5}: +1
SumsOfDoubles{.x = 3.5, .y = 2, .sum = 5.5}: +1
SumsOfDoubles{.x = 3.5, .y = 3.5, .sum = 7}: +1
InspectSimpleSum:
InspectSimpleSum{.x = 1, .total = 3}: +1
FilteredRelation:
FilteredRelation{.y = 2}: +1
FilteredRelation{.y = 3}: +1
FilteredRelation{.y = 5}: +1
FilteredRelation:
FilteredRelation{.y = 2}: -1
OutputInspectNot:
OutputInspectNot{.x = 4, .y = 4}: +1
OutputInspectNot:
OutputInspectNot{.x = 2, .y = 2}: +1
